modules/_labs/select/react/lib/Select.tsx

```diff
@@ -0,0 +1,787 @@
+    // * A disabled option was clicked (we ignore these clicks)
+    if (!this.isMenuInteractive() || optionProps.disabled) {
+      return;
+    }
+
+    this.setState({
+      focusedOptionIndex: index,
+      label: this.optionLabels[index],
+      selectedOptionIndex: index,
+    });
+
+    this.toggleMenu(false);
+    this.updateInput(this.optionValues[index]);
+  };
+
+  handleMenuBlur = (event: React.FocusEvent): void => {
+    this.toggleMenu(false);
+  };
+
+  handleKeyboardShortcuts = (event: React.KeyboardEvent): void => {
+    const childrenArray = React.Children.toArray(this.props.children) as React.ReactElement<
+      SelectOptionProps
+    >[];
+    const numOptions = childrenArray.length;
+    let isShortcut = false;
+    let nextFocusedIndex = 0;
+
+    // Check for type-ahead first
+    if (event.key.length === 1 && event.key.match(/\S/)) {
+      isShortcut = true;
+      this.handleKeyboardTypeAhead(event.key, numOptions);
+    } else {
+      switch (event.key) {
+        case 'ArrowUp':
+        case 'Up': // IE/Edge specific value
+        case 'ArrowDown':
+        case 'Down': // IE/Edge specific value
+          if (this.state.isMenuHidden) {
+            this.toggleMenu(true);
+          } else {
+            const direction = event.key === 'ArrowUp' || event.key === 'Up' ? -1 : 1;
+            isShortcut = true;
+            let nextIndex = this.state.focusedOptionIndex + direction;
+            while (
+              nextIndex < numOptions &&
+              nextIndex >= 0 &&
+              childrenArray[nextIndex].props.disabled
+            ) {
+              nextIndex += direction;
+            }
+            nextFocusedIndex =
+              nextIndex < 0 ? 0 : nextIndex >= numOptions ? numOptions - 1 : nextIndex;
+            this.setState({focusedOptionIndex: nextFocusedIndex});
+          }
+          break;
+
+        case 'Home':
+        case 'End':
+          isShortcut = true;
+          nextFocusedIndex = event.key === 'Home' ? 0 : numOptions - 1;
+          this.setState({focusedOptionIndex: nextFocusedIndex});
+          break;
+
+        case 'Spacebar':
+        case ' ':
+          isShortcut = true;
+          // If the user is in the middle of typing a string, treat
+          // space key as type-ahead rather than option selection
+          if (this.keysSoFar !== '') {
+            this.handleKeyboardTypeAhead(' ', numOptions);
+          } else {
+            this.handleOptionSelection(this.state.focusedOptionIndex);
+          }
+          break;
+
+        case 'Enter':
+          isShortcut = true;
+          this.handleOptionSelection(this.state.focusedOptionIndex);
+          break;
+
+        default:
+      }
+    }
+
+    if (isShortcut) {
+      event.stopPropagation();
+      event.preventDefault();
+    }
+  };
+
+  renderChildren = (
+    child: React.ReactElement<SelectOptionProps>,
+    index: number
+  ): React.ReactNode => {
+    return React.cloneElement(child, {
```

Comment:
Currently, `Select` injects several props into its `SelectOption` children via this `React.cloneElement` call: `error`, `focused`, `selected`, etc. These props are meant to be managed by the `Select` component and aren't intended to be explicitly passed into `SelectOption` (i.e., you should never use `<SelectOption focused={true} selected={true} ...>`).

For example, `selected` is a boolean which tells the `SelectOption` whether or not its selected -- the parent `Select` compares the `selectedOptionIndex` in its internal state against the index of the `SelectOption` and sets `selected` to `true` if they match.

Is there a different preferred method for implementing "private" props? @NicholasBoll mentioned using Context in an earlier comment (still investigating). I've also heard suggestions of render props (https://medium.com/better-programming/passing-data-to-props-children-in-react-5399baea0356).

Reply:
Render props would be my suggestion. The Tables 2.0 is going that direction. But also the way data is defined is different. The current API looks like this:

```tsx
<Select>
  <SelectItem />
  <SelectItem />
</Select>
```

While that is a nice clear API, it does have some issues with scaling. If you have a lot of items, you'll have to define them all. Possibly with a `map` function. The component will have to map over children again to add the extra props necessary. You'll have to make props like `selected`, `focused` optional so that Typescript doesn't complain to you. If we go for virtualization later (which we probably should for performance), this will still be a very expensive render because the render method is so large. Also the API is limited to only showing strings.

Another API would be slightly less obvious, but more flexible using a render prop:

```tsx
const options = [ { value: "United States", flag: "US" }, { value: "Ireland", flag: "IE" } ]
<Select
  options={options}
  renderItem={( itemProps, data ) => {
    <SelectOption {...itemProps} value={data.value}>
      <img src={`https://www.countryflags.io/${data.flag}/flat/32.png`} aria-hidden="true" /> {data.value}
    </SelectOption>
  }}
/>
```

This is a more flexible API where the rendering is more efficient and more obviously controlled by the `Select` component. We only need to process rendering of `SelectItems` only when the Select is opened. The render prop is called by the `Select` at the right time. Rendering will be more efficient. If the `Select` component optimizes by utilizing virtual rendering of items, our external API doesn't need to change and we don't waste time with expensive render methods. It is also more obvious that `options` can be extracted out of the render method entirely, allowing further optimizations by the `Select` component.

This API also makes it more obvious how dynamic `options`. If options comes back from the server, it will probably be an array anyways. We would just hand it to the component.

Reply:
@NicholasBoll @anicholls I've implemented the render prop (`renderOption`) in the latest changes if you'd like to check it out. I've included example code in the README as well as a `Custom Options` story.

It's a slightly modified version of Nicholas's proposed API -- I opted to have the render prop only handle the content of the `<SelectOption>` and not the `<SelectOption>` itself for simplicity.

It would be nice if we could use `renderOption` to render the actual select (which is actually a button) itself, but since it's a `<button>` we're limited in what we can include in its content. Something for a later PR.